File: apps/mobile/lib/features/patrol/domain/patrol.dart
enum PatrolSessionStatus { inProgress, completed }

PatrolSessionStatus _statusFromApi(Object? value) {
  return value == 'COMPLETED'
      ? PatrolSessionStatus.completed
      : PatrolSessionStatus.inProgress;
}

extension PatrolSessionStatusLabel on PatrolSessionStatus {
  String get label {
    switch (this) {
      case PatrolSessionStatus.inProgress:
        return 'Berlangsung';
      case PatrolSessionStatus.completed:
        return 'Selesai';
    }
  }
}

class PatrolCheckpoint {
  const PatrolCheckpoint({
    required this.id,
    required this.name,
    required this.displayOrder,
  });

  final String id;
  final String name;
  final int displayOrder;

  factory PatrolCheckpoint.fromJson(Map<String, dynamic> json) {
    return PatrolCheckpoint(
      id: json['id'] as String,
      name: json['name'] as String,
      displayOrder: (json['displayOrder'] as num).toInt(),
    );
  }
}

class PatrolScan {
  const PatrolScan({
    required this.checkpointId,
    required this.checkpointName,
    required this.scannedAt,
    required this.note,
  });

  final String checkpointId;
  final String checkpointName;
  final DateTime scannedAt;
  final String? note;

  factory PatrolScan.fromJson(Map<String, dynamic> json) {
    return PatrolScan(
      checkpointId: json['checkpointId'] as String,
      checkpointName: json['checkpointName'] as String,
      scannedAt: DateTime.parse(json['scannedAt'] as String),
      note: json['note'] as String?,
    );
  }
}

class PatrolSession {
  const PatrolSession({
    required this.id,
    required this.officerName,
    required this.status,
    required this.startedAt,
    required this.endedAt,
    required this.totalCheckpoints,
    required this.scans,
  });

  final String id;
  final String officerName;
  final PatrolSessionStatus status;
  final DateTime startedAt;
  final DateTime? endedAt;
  final int totalCheckpoints;
  final List<PatrolScan> scans;

  factory PatrolSession.fromJson(Map<String, dynamic> json) {
    final scans = json['scans'] as List<dynamic>? ?? const [];
    return PatrolSession(
      id: json['id'] as String,
      officerName: json['officerName'] as String,
      status: _statusFromApi(json['status']),
      startedAt: DateTime.parse(json['startedAt'] as String),
      endedAt: json['endedAt'] != null
          ? DateTime.parse(json['endedAt'] as String)
          : null,
      totalCheckpoints: (json['totalCheckpoints'] as num).toInt(),
      scans: scans
          .map((scan) => PatrolScan.fromJson(scan as Map<String, dynamic>))
          .toList(growable: false),
    );
  }
}

File: packages/contracts/src/patrol.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const patrolSessionStatusSchema = z.enum(["IN_PROGRESS", "COMPLETED"]);

export type PatrolSessionStatus = z.infer<typeof patrolSessionStatusSchema>;

export const scanCheckpointInputSchema = z
  .object({
    qrToken: z.string().trim().min(1).max(64),
    note: z.string().trim().max(500).optional(),
  })
  .strict();

export type ScanCheckpointInput = z.infer<typeof scanCheckpointInputSchema>;

export const patrolCheckpointSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1),
  displayOrder: z.number().int(),
});

export type PatrolCheckpoint = z.infer<typeof patrolCheckpointSchema>;

export const patrolScanSchema = z.object({
  checkpointId: z.string().uuid(),
  checkpointName: z.string().min(1),
  scannedAt: z.string().datetime({ offset: true }),
  note: z.string().nullable(),
});

export type PatrolScan = z.infer<typeof patrolScanSchema>;

export const patrolSessionSchema = z.object({
  id: z.string().uuid(),
  officerName: z.string().min(1),
  status: patrolSessionStatusSchema,
  startedAt: z.string().datetime({ offset: true }),
  endedAt: z.string().datetime({ offset: true }).nullable(),
  totalCheckpoints: z.number().int().nonnegative(),
  scans: z.array(patrolScanSchema),
});

export type PatrolSession = z.infer<typeof patrolSessionSchema>;

export const patrolCheckpointListResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(patrolCheckpointSchema) }),
);

export const patrolSessionResponseSchema = dataEnvelopeSchema(
  z.object({ session: patrolSessionSchema.nullable() }),
);

export const patrolSessionMutationResponseSchema = dataEnvelopeSchema(
  z.object({ session: patrolSessionSchema }),
);

export const patrolHistoryResponseSchema = dataEnvelopeSchema(
  z.object({ items: z.array(patrolSessionSchema) }),
);

export type PatrolCheckpointListResponse = z.infer<typeof patrolCheckpointListResponseSchema>;
export type PatrolSessionResponse = z.infer<typeof patrolSessionResponseSchema>;
export type PatrolSessionMutationResponse = z.infer<typeof patrolSessionMutationResponseSchema>;
export type PatrolHistoryResponse = z.infer<typeof patrolHistoryResponseSchema>;
